server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -403,7 +404,7 @@ public void onFailure(Exception e) {
 
                 @Override
                 public void onFailure(Exception e) {
-                    onComplete(exceptionToResult(e, primary, isDelete, version, result.getId()), context, updateResult);
+                    onMappingUpdateFailure(e, primary, isDelete, version, result, context, updateResult);
```

Comment:
I think it would be easier to understand this change and a better solution to let the mapping update itself simply not fail when the limit is exceeded, simply silently not adding the extra fields.

The `TransportAutoPutMapping` is already distinct from attempts to add the mapping manually and thus could be used to distinguish this during mapping merge. You'll likely need to add another merge reason like `MAPPING_AUTO_UPDATE`, that can handle this.

I think doing so will remove the need for the extra retries both here and above.

Reply:
A challenge I see with that approach is that wouldn't allow us to add the ignored fields to the `_ignored` metadata field. They'd be silently ignored with no indication that something went wrong with the document. This is where retrying the whole operation with an updated view of the mappings is useful.

IIUC, another downside is related to cases where multiple fields are added and part of the fields are within the limit while other parts exceed the limit. In that case, we wouldn't add any of the fields.

But maybe we can live with these limitations as it only affects the case when a concurrent operation adds fields so that we're closer to the field limit than we thought at the time we parsed the document.

Also, is it safe to perform the subsequent index request when the mapping update has been silently ignored? Is there any implicit assumption that the mapping update has succeeded? Or is it basically the same situation as if dynamic mapping had been disabled? From some initial testing, it seems to be the latter but maybe I'm missing something.

Reply:
Looking into `MAPPING_AUTO_UPDATE` a bit more, a challenge here is that we can't just always execute the behavior of "let the mapping update itself simply not fail when the limit is exceeded". We can only do that if the index has enabled `dynamic: true_until_limit`. But because `dynamic` can be set on any `object` field, this behavior isn't even consistent across the whole index but may be limited to just a specific path. We could make this an index setting instead, but we've decided against that (see https://github.com/elastic/elasticsearch/pull/96235#pullrequestreview-1520894321).

Reply:
After discussing this with @henningandersen and @javanna I gave this a shot in 5f3b6dc.

> A challenge I see with that approach is that wouldn't allow us to add the ignored fields to the _ignored metadata field. They'd be silently ignored with no indication that something went wrong with the document.

What I was missing is that the index request will be retried after a dynamic mapping update:

https://github.com/elastic/elasticsearch/blob/9f3b7ec3f516d27aa57ed3ce2689adc2cf983b1b/server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L394

Therefore, the document is parsed again so that the `_ignored` field gets populated.

Reply:
Some challenges with that approach, though:

As the mapping merge on master silently ignores fields above the limit, it may result in a noop where we skip updating the cluster state. Looks like the process that updates the mappings isn't prepared for a situation where the cluster state doesn't change and it will run into a timeout:

https://github.com/elastic/elasticsearch/blob/325f202cd819026152d3c74470e6e08ee4b09c78/server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L159

I believe this can even be an issue today when two threads are concurrently performing the same mapping update. The thread that comes second will yield a noop mapping update.

Once we fix that, another potential risk is that if we under-estimate the size of a mapper here (for example by not counting the additional shingle mappers in the `search_as_you_type` field type):

https://github.com/elastic/elasticsearch/blob/fac8dcb3b6aef77fa97d1e396f5fc5f2c06fc8ff/server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java#L319

In that case, I believe we could even get into some kind of infinite loop when the mapping is one field away from hitting the limit: While parsing the document, we think we can add the field (because we under-estimate the size) but when performing the dynamic mapping update, we realize the newly added `search_as_you_type` field turns out to add more than one mapper so that we silently ignore the whole field. After that, we retry the indexing request and the whole process repeats.

Reply:
I wonder if the preflight can detect this?

Regardless of approach I think we'd want to avoid contacting the master in such cases, i.e., if the only possible addition is 2 fields and only one is left (according to local knowledge), we should not contact master.

Reply:
I'll agree with both and I can look into implementing this. I just wanted to raise awareness that there's this risk now. Previously, both over or under estimating the size of a dynamic mapper didn't have much consequences. Now a bug in that area equals an infinite loop. But detecting noop mapping changes in the preflight and eagerly failing in that case already mitigates the risk somewhat.

@javanna could you or someone in your team have a look at the mapping merger changes to see if this is going in the right direction?

Reply:
Alright, in 1a735bd0e0bbece0ab56baecfdd230d64acc5b65, I added logic that skips the mapping update if the preflight indicates  a noop and added a safeguard against infinite retry loops.

Reply:
Due to a test failure in the 8.11.2 compatibility tests for dynamically mapping dense vectors, I've adjusted the infinite loop detection a bit in b3c8fda.

@henningandersen could you please have a look specifically at the changes in `server/src/main/java/org/elasticsearch/action/bulk` and `server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put` and approve if these look good to you?

Luca and team can then review the mapping changes.

Reply:
I've extracted the changes related to this in a separate PR: https://github.com/elastic/elasticsearch/pull/102915. I think that one can be merged independently which also reduces the size and complexity of this PR. @henningandersen could you or someone in your team  have a look? Thanks 🙏